akka-projection-jdbc/src/test/java/akka/projection/jdbc/JdbcProjectionTest.java

```diff
@@ -299,21 +320,20 @@ public void atLeastOnceShouldRestartFromPreviousOffset() {
                 projectionId,
                 sourceProvider(entityId),
                 jdbcSessionCreator,
-                // fail on forth offset
+                // fail on fourth offset
                 () -> concatHandler(str, offset -> offset == 4),
                 testKit.system())
             .withSaveOffset(1, Duration.ZERO);
 
-    try {
-      projectionTestKit.run(
-          projection,
-          () -> {
-            assertEquals("abc|def|ghi|", str.toString());
-          });
-      Assert.fail("Expected exception");
-    } catch (RuntimeException e) {
-      assertEquals(failMessage(4), e.getMessage());
-    }
+    projectionTestKit.runWithTestSink(
+        projection,
+        (probe) -> {
+          probe.request(2);
+          probe.expectNextN(2);
```

Comment:
I know why I only need to request 2 elements here.  I think it's because the at least once offset strategy creates a graph that performs an extra async operation somewhere. I could revert this change since we haven't logged any failures for this test. I updated it because the logic is nearly identical to what was changed in `exactlyOnceShouldRestartFromPreviousOffset`.

Reply:
maybe useful to add a small comment explaining why we request 2, but get `abc|def|ghi`

Reply:
It's complicated, so I'm not sure a comment would suffice.  I dug into this more and narrowed down the issue to additional demand being generated in the at-least-once use case for the JDBC handler.  When we save the offset for the at-most-once and at-least-once use cases we asynchronously save the offset within a `mapAsync(parallelism = 1)` stage. Since this stage is downstream from the processing stage it generates extra demand sort of like a buffer.

In the exactly-once case we are performing a blocking call to save the offset within the same transaction as the `DBIO` returned by the envelope handler, so there's no additional `mapAsync` or other buffering/grouping occurring.

I talked with @ignasi35 because I thought this originally had to do with how the exacty-once use case is instrumented for metrics, but now I know it's due async saving of offsets as explained above.  An idea we had is to provide a utility in the test kit to expect a certain number of elements from the source provider so we know exactly how many elements enter the projection impl part of the graph. Then this test would use the utility to tell the source provider only to pass _n_ elements downstream and then do await/assert until the projection state is reached. This will work in many projections use cases, but this issue is just a consequence of how more complex async graphs work. At runtime I don't think there's any problem, but the nuances are going to be difficult to explain at the testing level.

/cc @patriknw since I've had similar discussions about this before.

Reply:
Interesting. 

Would be possible to add such an utility without exposing it to users? I mean, could we wrap the source in such a way that we only let an element pass when the previous have been processed?

Reply:
For the record, I have noticed this before when writing the tests and I think I saw @ignasi35 commenting on that as well, but I didn't look further to understand why this was happening. At the time I thought that it was just because of some buffering, I didn't notice that it only happens when using `mapAsync`.

Reply:
I experimented with adding a probe to the `TestSourceProvider`, but it would mean somehow propagating the mat value (the probe) through the whole graph to the projection impl to be materialized, since it would need to be materialized only once with the projection stream itself.

Instead I'm going to try and use a `Source.queue` to offer just the number of elements I want to at various stages of the test. This should have the same effect as using a probe and might be more intuitive.

Reply:
Actually, that won't work either because I still need to propagate the queue mat value through the stream.  I'll create a case instead that describes the issue, and I'll put in a comment as you originally proposed.

Reply:
https://github.com/akka/akka-projection/issues/462